add comprehensive runtime library configuration support

shanedsnyder opened the pull request:
This PR encompasses a lot of changes to darshan-runtime library to support more comprehensive configuration of Darshan instrumentation behavior, memory usage, etc. It includes the following changes:

- Ability to specify a user configuration file containing different settings to configure Darshan library with. This config file supports many of the settings traditionally provided via env vars, but also more detailed settings described below. 
- Refactoring of darshan-runtime to set an initial configuration, to extract user-specific settings from the env or a given file, and to apply settings throughout runtime (as modules and records are registered with Darshan). Special care is taken to ensure env vars override any settings given in config files.
- Changes to Darshan's module API to allow modules to be controlled by user settings.
- Refactoring of DXT to more strictly adhere to Darshan's module API so that it can be more easily controlled by Darshan config settings.
- Add an ability to print Darshan's configuration to `stderr` to sanity check user settings were applied correctly (via `DARSHAN_DUMP_CONFIG` env var).
- darshan-runtime documentation changes to describe new configuration hooks in detail.

New settings allow users to enable/disable any module at runtime:

```
# enable non-default DXT modules
MOD_ENABLE        DXT_POSIX,DXT_MPIIO
# disable always-on POSIX module
MOD_DISABLE       POSIX
```

Users have more control over how many records Darshan allocates for each module, to avoid hitting maximums:

```
# bump up default record allocations for POSIX and STDIO modules
MAX_RECORDS       4096            POSIX,STDIO
# Note that '*' wildcard can be used to apply to all modules
MAX_RECORDS       2000               *
```

Users can selectively exclude/include instrumentation for specific record names:

```
# exclude any record names starting with "/home" (i.e., to home directory)
NAME_EXCLUDE      ^/home              *
# overriding include for files with ".out" suffixes
NAME_INCLUDE      .out$                   *
```

Users can selectively disable/enable Darshan using regular expression matching on application names:

```
# exclude Darshan instrumentation for binaries we don't care about
APP_EXCLUDE       (git|ls|sed)
```

Users can also selectively enable/disable Darshan instrumentation on specific ranks:

```
# exclude instrumentation on all ranks
RANK_EXCLUDE     0:
# re-enable instrumentatio on ranks 0-9
RANK_INCLUDE     0:9
```

The snippets above all are provided as they would be in a config file, but the general syntax is mostly the same for environment variables. Environment variables are additionally prefixed with `DARSHAN_` and there is no env var support for NAME_EXCLUDE/NAME_INCLUDE/MAX_RECORDS. Best place to learn more details about each of these settings is the darshan-runtime docs that have been updated.

Replaces stale #405.

shanedsnyder commented:
Note that I have punted on providing a default Darshan config file that installs with the library. That's probably a straightforward addition at some point if we feel the need, but seemed like overkill to install something that reproduces the default values already in our code (which would probably hang around anyway) -- most of it's value would probably just be a template for users to copy, but that can just go in our documentation.

shanedsnyder commented:
It's pretty verbose (there's a lot of settings!), but here's what you get when you dump the Darshan config to stderr for now for a sample config I've been messing around with on my laptop:
```
##########################
##### DARSHAN CONFIG #####
##########################
# MODMEM = 8 MiB
# NAMEMEM = 2048 KiB
# MEM_ALIGNMENT = 8 bytes
# JOBID = NONE
# LOGHINTS = romio_no_indep_rw=true;cb_nodes=4
# LOGPATH = /home/shane/software/darshan/darshan-logs
# LOGPATH_BYENV = NONE
# MMAP_LOGPATH = /tmp
# EXCLUDE_DIRS = /etc/,/dev/,/usr/,/bin/,/boot/,/lib/,/opt/,/sbin/,/sys/,/proc/,/var/
# APP_EXCLUDE = (git|ls|sed)
# APP_INCLUDE = NONE
# RANK_EXCLUDE = NONE
# RANK_INCLUDE = NONE
# POSIX MODULE CONFIG:
#      - ENABLED
#      - MAX_RECORDS = 4096
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# MPI-IO MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# H5F MODULE CONFIG:
#      - DISABLED
# H5D MODULE CONFIG:
#      - DISABLED
# PNETCDF MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# BG/Q MODULE CONFIG:
#      - DISABLED
# LUSTRE MODULE CONFIG:
#      - DISABLED
# STDIO MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# DXT_POSIX MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# DXT_MPIIO MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
# MDHIM MODULE CONFIG:
#      - DISABLED
# APXC MODULE CONFIG:
#      - DISABLED
# APMPI MODULE CONFIG:
#      - DISABLED
# HEATMAP MODULE CONFIG:
#      - ENABLED
#      - NAME_EXCLUDE = ^/home
#      - NAME_INCLUDE = .out$
##########################
### END DARSHAN CONFIG ###
##########################

```
Provides details on a lot of traditional settings, but also shows what modules are enabled, module-specific settings, etc.

carns commented:
Thanks @shanedsnyder.  I haven't looked at the code yet, but a couple of quick feedback points:

- I'm getting some new warnings when compiling with `-g -Wall`, can we clean those up?
- The build is failing with --enable-apmpi-mod.  Are we going to need to coordinate some changes in autoperf when this PR lands?

carns commented on darshan-runtime/lib/darshan-core.c:
```diff
@@ -759,39 +756,1085 @@ void darshan_core_shutdown(int write_log)
+        darshan_core_fprintf(stderr, "darshan:job_write\t%d\t%f\n", nprocs, job_tm);
+        darshan_core_fprintf(stderr, "darshan:hash_write\t%d\t%f\n", nprocs, rec_tm);
+        darshan_core_fprintf(stderr, "darshan:header_write\t%d\t%f\n", nprocs, header_tm);
+        for(i = 0; i < DARSHAN_MAX_MODS; i++)
```
DARSHAN_MAX_MODS is not safe for this loop, because the i variable is being used as an index into darshan_module_names[], which is declared with an x macro and only has DARSHAN_KNOWN_MODULE_COUNT entries in it.  Might need to audit and see if there are any other loops like this where this could be a problem.

Or maybe a safer solution would be a macro that looks up the name of a module based on it's id and returns a null string or "UNKNOWN" if the id > KNOWN_MODULE_COUNT.  That way we wouldn't have to worry about manually indexing for the x macro arrays that have a smaller size.

shanedsnyder replied:
I can open an issue to look more at this. There do appear to be 1 or more spots in the runtime code that could be problematic, but I don't see any introduced in this PR. Actually, in the `dump_config()` routine, I'm surprised I remembered to do the right thing with `DARSHAN_KNOWN_MODULE_COUNT`.

shanedsnyder replied:
Opened #712 to track this

carns reviewed (changes requested):
I put comments on a few things (some are just questions, a few are requests, one is probably best handled as a separate issue).  At a high level we have a few -Wall warnings to fix and a plan for how to coordinate API updates with AutoPerf.

shanedsnyder commented:
> Thanks @shanedsnyder. I haven't looked at the code yet, but a couple of quick feedback points:
> 
>     * I'm getting some new warnings when compiling with `-g -Wall`, can we clean those up?
> 

Good call, I missed those but I will clean them up.

>     * The build is failing with --enable-apmpi-mod.  Are we going to need to coordinate some changes in autoperf when this PR lands?

That's right. I can put together a PR for AutoPerf repo and we can make sure to get that merged right after this. The changeset should be pretty modest for those modules.

shanedsnyder commented on darshan-runtime/doc/darshan-runtime.txt:
```diff
@@ -478,6 +446,147 @@ git submodule update --init
+specified in both the environment and via a config file, the environment
+settings will take precedence.
+
+Users may specify the config file location using the `DARSHAN_CONFIG_PATH`
```
Just in case you missed this in my back-to-back-to-back comments from above: https://github.com/darshan-hpc/darshan/pull/703#issuecomment-1088074029

I opted for not providing a default config file installed with Darshan for the time being just because I wasn't sure the additional complexity was worth the effort, but we can revisit if I'm missing an obvious benefit. If we added a default config file, it would really just be duplicating hard-coded settings in the code base, so not sure what the benefits are for that additional wrinkle beyond just providing an example users could build off of?

carns replied:
Whoops yeah, I missed it.  Let's stick with that.  In the documentation should we maybe suggest (for end users) that if they want to provide their own config file, that they check to see if the DARSHAN_CONFIG_PATH env var is already in use by the facility (like if the ALCF provided one in their site darshan module), and if so copy that one to use as a starting point?  Just a safety check to make sure that a user doesn't accidentally alter unrelated things by replacing a site config.

shanedsnyder replied:
I added a note about this to the docs in 3442f42. I think the NOTE could be in a better spot, but I'm still planning another pass over the docs and will fix that up as part of your other feedback.

shanedsnyder commented on darshan-runtime/lib/darshan-core.c:
```diff
@@ -759,39 +756,1085 @@ void darshan_core_shutdown(int write_log)
+
+/* *********************************** */
+
+static void darshan_init_config(struct darshan_core_runtime *core)
```
That's actually a good idea, I should have thought of that since you're right that darshan-core source is getting massive. These would be pretty easy to break off into their own source file that operates on Darshan config structures that are populated for use by darshan-core.

shanedsnyder commented on darshan-runtime/lib/darshan-lustre.c:
```diff
@@ -228,19 +231,23 @@ static void lustre_runtime_initialize()
 
     /* register the lustre module with darshan-core */
-    darshan_core_register_module(
+    /* assume each file uses 64 OSTs */
```
That's right, the Lustre module needs to provide some indication of how much memory it needs at runtime, but it has dynamic-length records, so just assumes that all records will require 64 OSTs, then requests the default number of Darshan records (1024) with that assumption. The Lustre module has always been coded this way, just that it might be more obvious now given the change to provide both `nrecs` and `sizeof(rec)` to `module_register()`. I'll add comments to make sure this is clear, but you're right that the Lustre module could run out of memory with more or less records than what it requested at init time.

shanedsnyder commented on darshan-runtime/doc/darshan-runtime.txt:
```diff
@@ -478,6 +446,147 @@ git submodule update --init
+Users may specify the config file location using the `DARSHAN_CONFIG_PATH`
+environment variable. The config file allows users to specify the following
+settings, one per-line, with setting names and values separated by any whitespace.
+Many of the settings directly match settings defined in section 
```
That was kind of the intent, there are only a handful of exceptions right now, maybe you can comment on whether these exceptions make sense:

- `MAX_RECORDS` and `NAME_EXCLUDE`/`NAME_INCLUDE`: Not supported as env variable variants because these settings require a value and a list of modules to apply them to, which is complicated to express as an env var. You'd either have to come up with some wonky formatting in a single string, or you'd have to define independent env vars for each module which I don't think we want.
- `DXT_ENABLE_IO_TRACE`: Not supported in config file for a couple of reasons. 1) `MODULE_ENABLE` settings can accomplish the same thing, so decided it might be simplest to only honor `DXT_ENABLE_IO_TRACE` as an env var as a convenience to users who have gotten used to it that way. 2) This env var isn't prefixed with `DARSHAN`, which makes it an oddball to try to support in the config file in this 1-1 way (chop off the `DARSHAN_` part of the env var to get the config file setting).
- `DARSHAN_EXCLUDE_DIRS`: Not supported in config file for similar reasons to `DXT_ENABLE_IO_TRACE`, in that the config file has a new way of doing the same thing via `NAME_EXCLUDE`. We could honor this in the config file, I just think there's a tradeoff to being consistent and not advertising multiple ways for users to do the same thing (while still honoring old env vars that users may have gotten used to).

So, first is because of complexity, and the latter 2 are related to deprecation.

Beyond that, there are some smaller examples of discrepancies with the following only available as env vars:
- `DARSHAN_DUMP_CONFIG`:
- `DARSHAN_INTERNAL_TIMING` 
- `DARSHAN_DISABLE_SHARED_REDUCTION`
- `DARSHAN_DISABLE`
- `DARSHAN_LOGNAME`

I suppose we could offer config file variants for the first 3 if we wanted, they just kind of struck me as small things that could just stick to the environment only. The last 2 don't really seem appropriate for a config file.

shanedsnyder replied:
I agree with the comments in general though, I'll take another pass over this part of the documentation and try to unify as much of this as possible instead of repeating a lot of the things in 2 different sections. I'll wait and see your feedback on discrepancies before hacking up any changes to the code though.

shanedsnyder replied:
Okay, I think these changes are reflected in my recent commits. It'd probably be most helpful if you can scan the updated docs to see if the new section and table I came up with are sensible. There are some changes to runtime code to support more settings in the config file approach (specifically, `DISABLE_SHARED_REDUCTION`, `DUMP_CONFIG`, and `INTERNAL_TIMING`).

I also tried to make sure to be more consistent in how users can set settings depending on whether they use env vars or a config file. Specifically, I made sure to be consistent that users can provide CSVs of regular expressions for all settings, regardless of whether they are using env vars or the config file -- obviously, that is a necessity for env vars (you can't set them multiple times like you can repeatedly specify config file settings), but I wasn't supporting that in the config file approach. Now, the config file allows you to specify these parameters using multiple lines, or using comma-separated values, if that makes sense.

carns commented on darshan-runtime/lib/darshan-core.c:
```diff
@@ -759,39 +756,1085 @@ void darshan_core_shutdown(int write_log)
+        token = strtok(string, ",");
+        while(token != NULL)
+        {
+            regex = malloc(sizeof(*regex));
```
ohhh, I see.  Thanks for the explanation.

shanedsnyder commented on darshan-runtime/lib/darshan-core.c:
```diff
@@ -759,39 +756,1085 @@ void darshan_core_shutdown(int write_log)
+    return;
+}
+
+#define DARSHAN_PARSE_NUMBER_FROM_STR(__str, __cast, __out, __success) do { \
```
Done in 3320912

shanedsnyder commented:
Think I'm almost done with the first round of comments. I still need to:

- get a PR into the AutoPerf repo that incorporates changes to the module API
- iterate on config code/docs to try to unify as much as we can across environment and file configuration (from this discussion: https://github.com/darshan-hpc/darshan/pull/703#discussion_r842726506)

shanedsnyder commented:
PR for AutoPerf is available now: https://github.com/argonne-lcf/autoperf/pull/14

We should be able to have that merged once we're satisfied here. (I still have one more change mentioned above to work on).

shanedsnyder commented:
I think I'm all caught up on reviewer comments after these commits and ready for a re-review. Beyond the unresolved issue above, these commits:

- added a new `--with-name-mem` configure arg to mirror `--with-mod-mem` one
- added better corner-case handling for extracting rank ranges (for RANK_EXCLUDE/INCLUDE settings)
- fixed a configure typo for `--with-mem-align`
- fixed general typos in runtime docs

shanedsnyder commented:
Perfect, thanks for having a look! And I'm not surprised about the table not rendering properly, it was a bit of a battle to get it working in the HTML even. I had similar thoughts about readthedocs and didn't spend a whole lot of time trying to smooth it out, but hoping we make time to make the move at some point in the near future.
